src/coll_score/ucc_coll_score.c

```diff
@@ -728,83 +851,117 @@ static ucc_status_t ucc_coll_score_update_one(ucc_list_link_t *dest,
             new       = MSG_RANGE_DUP(rd);
             new->end  = rs->start;
             rd->start = rs->start;
-            ucc_list_insert_before(d, &new->list_elem);
+            ucc_list_insert_before(d, &new->super.list_elem);
         } else if (rd->start > rs->start) {
-            if (rs->init) {
+            if (rs->super.init) {
                 new = MSG_RANGE_DUP(rs);
-                if (new->score == UCC_SCORE_INVALID) {
-                    new->score = default_score;
+                if (new->super.score == UCC_SCORE_INVALID) {
+                    new->super.score = default_score;
                 }
                 new->end = rd->start;
-                ucc_list_insert_before(d, &new->list_elem);
+                ucc_list_insert_before(d, &new->super.list_elem);
             }
             rs->start = rd->start;
         } else {
             /* same start */
             if (rs->end > rd->end) {
-                if (UCC_SCORE_INVALID != rs->score) {
-                    rd->score = rs->score;
+                if (UCC_SCORE_INVALID != rs->super.score) {
+                    rd->super.score = rs->super.score;
                 }
-                if (rs->init) {
-                    rd->init = rs->init;
-                    rd->team = rs->team;
+                if (rs->super.init) {
+                    if (rs->super.init != rd->super.init) {
+                        /* User setting overrides existing init fn. Save it as a fallback */
+                        status =
+                            ucc_fallback_alloc(rd->super.score, rd->super.init,
+                                               rd->super.team, &fb);
+                        if (ucc_unlikely(UCC_OK != status)) {
+                            goto out;
+                        }
+                        ucc_fallback_insert(&rd->fallback, fb);
+                    }
+                    rd->super.init = rs->super.init;
+                    rd->super.team = rs->super.team;
                 }
                 rs->start = rd->end;
                 d         = d->next;
             } else if (rs->end < rd->end) {
                 new      = MSG_RANGE_DUP(rd);
                 new->end = rs->end;
-                if (UCC_SCORE_INVALID != rs->score) {
-                    new->score = rs->score;
+                if (UCC_SCORE_INVALID != rs->super.score) {
+                    new->super.score = rs->super.score;
                 }
-                if (rs->init) {
-                    new->init = rs->init;
-                    new->team = rs->team;
+                if (rs->super.init) {
+                    if (rs->super.init != rd->super.init) {
```

Comment:
seems Line  894-905 is duplicated many times. can we merge ucc_fallback_alloc and ucc_fallback_insert into a single functions?

Reply:
done